File: week1/boj1799.py
def search(yi):
    global cnt
    global mx
    for y in range(yi,N):
        for x in range(N):
            if not bl[y][x] or d1[x+y] or d2[x-y+N-1]: continue
            d1[x+y] = d2[x-y+N-1] = True
            cnt += 1
            search(y)
            d1[x+y] = d2[x-y+N-1] = False
            cnt -= 1
    if cnt > mx: mx = cnt

N_MAX = 11
d1 = [False for _ in range(N_MAX*2)]
d2 = d1[:]
bl = []
cnt = mx = 0
N = int(input())
for Y in range(N):
    bl.append(list(map(int,input().split())))
search(0)
print(mx)

File: week1/boj1799.cpp
#include <cstdio>
using namespace std;

const int N_MAX = 11;
bool bl[N_MAX][N_MAX], d1[N_MAX*2], d2[N_MAX*2];
int cnt, mx, N;

void search(int yi)
{
    for (int y = yi; y < N; y++) {
        for (int x = 0; x < N; x++) {
            if (!bl[y][x] || d1[x+y] || d2[x-y+N-1]) continue;
            d1[x+y] = d2[x-y+N-1] = 1;
            cnt++;
            search(y);
            d1[x+y] = d2[x-y+N-1] = 0;
            cnt--;
        }
    }
    if (cnt > mx) mx = cnt;
}

int main()
{
    scanf("%d",&N);
    for (int y = 0; y < N; y++) {
        for (int x = 0; x < N; x++) {
            scanf("%d",&bl[y][x]);
        }
    }
    search(0);
    printf("%d\n",mx);

    return 0;
}

